ScoutSuite/providers/azure/resources/network/network_security_groups.py

```diff
@@ -0,0 +1,129 @@
+
+            source_port_ranges = self._merge_prefixes_or_ports(sr.source_port_range, sr.source_port_ranges)
+            security_rule_dict['source_port_ranges'] = source_port_ranges
+            security_rule_dict['source_ports'] = self._parse_ports(source_port_ranges)
+
+            destination_address_prefixes = self._merge_prefixes_or_ports(sr.destination_address_prefix,
+                                                                         sr.destination_address_prefixes)
+            security_rule_dict['destination_address_prefixes'] = destination_address_prefixes
+
+            destination_port_ranges = self._merge_prefixes_or_ports(sr.destination_port_range,
+                                                                    sr.destination_port_ranges)
+            security_rule_dict['destination_port_ranges'] = destination_port_ranges
+            security_rule_dict['destination_ports'] = self._parse_ports(destination_port_ranges)
+
+            security_rule_dict['etag'] = sr.etag
+
+            security_rules[security_rule_dict['id']] = security_rule_dict
+
+        return security_rules
+
+    def _parse_ports(self, port_ranges):
+        ports = set()
+        for pr in port_ranges:
+            if pr == "*":
+                for p in range(0, 65535 + 1):
+                    ports.add(p)
+                break
+            elif "-" in pr:
+                lower, upper = pr.split("-")
+                for p in range(int(lower), int(upper) + 1):
+                    ports.add(p)
+            else:
+                ports.add(int(pr))
+        ports = list(ports)
+        ports.sort()
+        return ports
+
+    def _parse_exposed_ports(self, network_security_group):
+        exposed_ports = set()
+
+        # Sort by priority.
+        rules = network_security_group.default_security_rules + network_security_group.security_rules
+        rules.sort(key=lambda x: x.priority, reverse=True)
+
+        for sr in rules:
+            if sr.direction == "Inbound" and (sr.source_address_prefix == "*"
+                                              or sr.source_address_prefix == "Internet"):
+                port_ranges = self._merge_prefixes_or_ports(sr.destination_port_range,
+                                                                     sr.destination_port_ranges)
+                ports = self._parse_ports(port_ranges)
+                if sr.access == "Allow":
+                    for p in ports:
+                        exposed_ports.add(p)
+                else:
+                    for p in ports:
+                        exposed_ports.discard(p)
+        exposed_ports = list(exposed_ports)
+        exposed_ports.sort()
+        return exposed_ports
+
+    def _merge_prefixes_or_ports(self, port_range, port_ranges):
+        port_ranges = port_ranges if port_ranges else []
+        if port_range:
+            port_ranges.append(port_range)
+        return port_ranges
+
+    def _format_ports(self, ports):
```

Comment:
I feel like it would be worth writing one or two unit tests for this method, `_parse_exposed_ports` and `_parse_ports`. It should be pretty quick since they won't require any mocking or setup. It's a bit hard to see what the methods do, having unit tests would be a good way of documenting them.

Reply:
Would you mind if I create an issue to do that and postpone it? I didn't write the `_format_ports` myself and this PR is just about moving the current code to the new architecture...

Reply:
Of course 👍 !

Reply:
Thanks, done! (#255 )